dali/pipeline/operators/arithmetic/arithmetic.h

```diff
@@ -0,0 +1,249 @@
+#define DALI_PIPELINE_OPERATORS_ARITHMETIC_ARITHMETIC_H_
+
+#include <memory>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "dali/core/static_switch.h"
+#include "dali/kernels/tensor_shape.h"
+#include "dali/kernels/tensor_shape_print.h"
+#include "dali/kernels/type_tag.h"
+#include "dali/pipeline/operators/arithmetic/arithmetic_meta.h"
+#include "dali/pipeline/operators/arithmetic/expression_impl_factory.h"
+#include "dali/pipeline/operators/operator.h"
+
+namespace dali {
+
+/**
+ * @brief Cover every sample fully.
+ */
+inline std::tuple<std::vector<TileDesc>, std::vector<TileRange>> GetCover(
+    const kernels::TensorListShape<> &shape, int num_tasks) {
+  std::vector<TileDesc> descs;
+  std::vector<TileRange> ranges(num_tasks, {0, 0});
+  int samples_per_task = shape.num_samples() / num_tasks;
+  if (!samples_per_task)
+    samples_per_task = shape.num_samples();
+  int previous_idx = -1;
+  for (int sample_idx = 0; sample_idx < shape.num_samples(); sample_idx++) {
+    int task_idx = sample_idx / samples_per_task;
+    descs.push_back({sample_idx, 0, task_idx, shape[sample_idx].num_elements(),
+                     shape[sample_idx].num_elements()});
+    if (task_idx != previous_idx) {
+      ranges[task_idx].begin = descs.size() - 1;
+      ranges[task_idx].end = descs.size();
+    } else {
+      ranges[task_idx].end = descs.size();
+    }
+    previous_idx = task_idx;
+  }
+  return std::make_tuple(descs, ranges);
+}
+
+/**
+ * @brief Divide the shape into groups of linear extents, separated evenly between threads
+ */
+inline std::tuple<std::vector<TileDesc>, std::vector<TileRange>> GetTiledCover(
+    const kernels::TensorListShape<> &shape, int extent, int num_tasks) {
+  Index total_elements = shape.num_elements();
+  std::vector<TileDesc> descs;
+  std::vector<TileRange> ranges(num_tasks, {0, 0});
+  Index elements_per_thread = total_elements / num_tasks;
+  // One extent covers all
+  if (total_elements < extent || elements_per_thread == 0) {
+    return GetCover(shape, num_tasks);
+  }
+  int task_idx = 0;
+  int previous_idx = -1;
+  std::vector<Index> covered_by_thread(num_tasks, 0);
+  for (int sample_idx = 0; sample_idx < shape.num_samples(); sample_idx++) {
+    int extent_idx = 0;
+    for (Index covered = 0; covered < shape[sample_idx].num_elements(); covered += extent) {
+      auto actually_covered =
+          std::min(static_cast<Index>(extent), shape[sample_idx].num_elements() - covered);
+      descs.push_back({sample_idx, extent_idx, task_idx, actually_covered, extent});
+      // We either cover a full extent, of what is left in the sample
+      covered_by_thread[task_idx] += actually_covered;
+      // Update range information
+      if (task_idx != previous_idx) {
+        ranges[task_idx].begin = descs.size() - 1;
+        ranges[task_idx].end = descs.size();
+      } else {
+        ranges[task_idx].end = descs.size();
+      }
+      previous_idx = task_idx;
```

Comment:
I'm probably missing the point. Why we want to have threads and extent inside them? Do we need `extent` at all?

Reply:
So we have 3 units of work. I will add this as a comment in a moment:
* thread - fixed number of threads that can pick up tasks added to thread pool
* task - task in this case consists of a set of tiles processed by thread.
* tile  - describes a portion of linear buffer, we try to split the amount of work evenly into tasks.

There should be more tasks than threads - so they can be distributed better.
I group tiles into tasks, so we reduce the switching of what the thread is executing.
The smaller tiles will help with when we have intermediate results in the expression - you need buffers as big as the tile size for that.

Reply:
So task is split to tile to limit the size of the intermediate buffers, right?

Reply:
Basically yes. And when you have a big batch of small samples, than one job/task in a thread pool will be able to process several samples at a time.